src/html5/parser.rs

```diff
@@ -268,1028 +274,1248 @@ impl<'stream> Html5Parser<'stream> {
+    fn process_foreign_content(&mut self) {
+        let mut handle_as_script_endtag = false;
+
+        match &self.current_token.clone() {
+            Token::TextToken { .. } if self.current_token.is_null() => {
+                self.parse_error("null character not allowed in foreign content");
+                self.insert_text_element(&Token::TextToken {
+                    value: CHAR_REPLACEMENT.to_string(),
+                });
+            }
+            Token::TextToken { .. } if self.current_token.is_empty_or_white() => {
+                self.insert_text_element(&self.current_token.clone());
             }
+            Token::TextToken { .. } => {
+                self.insert_text_element(&self.current_token.clone());
 
-            match self.insertion_mode {
-                InsertionMode::Initial => {
-                    let mut anything_else = false;
+                self.frameset_ok = false;
+            }
+            Token::CommentToken { .. } => {
+                self.insert_comment_element(&self.current_token.clone(), None);
+            }
+            Token::DocTypeToken { .. } => {
+                self.parse_error("doctype not allowed in foreign content");
+                // ignore token
+            }
+            Token::StartTagToken { name, .. }
+                if name == "b"
+                    || name == "big"
+                    || name == "blockquote"
+                    || name == "body"
+                    || name == "br"
+                    || name == "center"
+                    || name == "code"
+                    || name == "dd"
+                    || name == "div"
+                    || name == "dl"
+                    || name == "dt"
+                    || name == "em"
+                    || name == "embed"
+                    || name == "h1"
+                    || name == "h2"
+                    || name == "h3"
+                    || name == "h4"
+                    || name == "h5"
+                    || name == "h6"
+                    || name == "head"
+                    || name == "hr"
+                    || name == "i"
+                    || name == "img"
+                    || name == "li"
+                    || name == "listing"
+                    || name == "menu"
+                    || name == "meta"
+                    || name == "nobr"
+                    || name == "ol"
+                    || name == "p"
+                    || name == "pre"
+                    || name == "ruby"
+                    || name == "s"
+                    || name == "small"
+                    || name == "span"
+                    || name == "strong"
+                    || name == "strike"
+                    || name == "sub"
+                    || name == "sup"
+                    || name == "table"
+                    || name == "tt"
+                    || name == "u"
+                    || name == "ul"
+                    || name == "var" =>
```

Comment:
)We should probably be using a match statement instead of an if statement here.)